foundryup/src/process.rs

```diff
@@ -0,0 +1,342 @@
+        f.pad("a global process has already been set")
+    }
+}
+
+impl std::error::Error for SetGlobalProcessError {}
+
+#[derive(Clone, Debug)]
+pub struct Process {
+    process: Arc<dyn Processor>,
+}
+
+impl Process {
+    /// Returns a `Process` that forwards to the given [`Processor`].
+    pub fn new<S>(process: S) -> Self
+    where
+        S: Processor,
+    {
+        Self { process: Arc::new(process) }
+    }
+}
+
+impl Default for Process {
+    fn default() -> Self {
+        Self { process: Arc::new(DefaultProcess::default()) }
+    }
+}
+
+impl Deref for Process {
+    type Target = dyn Processor;
+
+    fn deref(&self) -> &Self::Target {
+        &*self.process
+    }
+}
+
+impl dyn Processor {
+    /// Returns `true` if this `Processor` is the same type as `T`.
+    pub fn is<T: Any>(&self) -> bool {
+        self.downcast_ref::<T>().is_some()
+    }
+
+    /// Returns some reference to this `Processor` value if it is of type `T`,
+    /// or `None` if it isn't.
+    pub fn downcast_ref<T: Any>(&self) -> Option<&T> {
+        unsafe {
+            let raw = self.downcast_raw(TypeId::of::<T>())?;
+            Some(&*(raw.cast().as_ptr()))
+        }
+    }
+}
+
+/// The process state of a thread.
+struct State {
+    /// This thread's current processor.
+    current: RefCell<Process>,
+}
+
+impl State {
+    /// Replaces the current  processor on this thread with the provided
+    /// processor.
+    ///
+    /// Dropping the returned `ScopeGuard` will reset the current processor to
+    /// the previous value.
+    #[inline]
+    fn set_current(new_process: Process) -> ScopeGuard {
+        let prior = CURRENT_STATE.try_with(|state| state.current.replace(new_process)).ok();
+        EXISTS.store(true, Ordering::Release);
+        CURRENT_COUNT.fetch_add(1, Ordering::Release);
+        ScopeGuard(prior)
+    }
+}
+
+/// A guard that resets the current processor to the prior
+/// current processor when dropped.
+#[derive(Debug)]
+pub struct ScopeGuard(Option<Process>);
+
+impl Drop for ScopeGuard {
+    #[inline]
+    fn drop(&mut self) {
+        CURRENT_COUNT.fetch_sub(1, Ordering::Release);
+        if let Some(process) = self.0.take() {
+            // Replace the processor and then drop the old one outside
+            // of the thread-local context.
+            let prev = CURRENT_STATE.try_with(|state| state.current.replace(process));
+            drop(prev)
+        }
+    }
+}
+
+/// The standard `Process` impl
+#[derive(Copy, Clone, Debug, Default)]
+pub struct DefaultProcess(());
+
+impl Processor for DefaultProcess {}
+
+#[cfg(test)]
+mod tests {
+    use super::*;
+
+    #[test]
+    fn current_processor_works() {
+        #[derive(Debug)]
+        struct TestProcessor;
+        impl Processor for TestProcessor {}
+
+        with(&Process::new(TestProcessor), || {
+            with_default(|processor| assert!(processor.is::<TestProcessor>()))
+        });
+    }
+}
```

Comment:
this whole file seems a little crazy although i recall seeing it before

Reply:
I can definitely trim this a bit and remove what's not needed